_posts/2020-03-02-reduce-boilerplate-with-records.md

```diff
@@ -0,0 +1,212 @@
+</project>
+```
+
+Note! The compiler needs a flag "--enable-preview", otherwise you'll get a compilation failure saying: "records are a preview feature and are disabled by default".
+
+## Implement a sample application
+
+I created a ```test-java-14/src/test/java/fi/solita/BloggingAppTest.java``` and ran it with ```mvn test```.
+This is what the whole thing looks like:
+
+```
+package fi.solita;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import org.junit.Test;
+import java.util.List;
+
+
+record Post(long blogId, String author, String title, String content) {}
+record Blog(long id, List<Post> posts) {}
+
+
+public class BloggingAppTest {
+    private Gson gson = new GsonBuilder().setPrettyPrinting().create();
+
+    @Test
+    public void records_are_serialized_to_json() {
+        var blog = createTestBlog();
+        String json = gson.toJson(blog);
+        System.out.println(json);
+
+        // outputs:
+        /*
+        {
+          "id": 1,
+          "posts": [
+            {
+              "blogId": 1,
+              "author": "Janne",
+              "title": "This is title",
+              "content": "This is content"
+            }
+          ]
+        }
+         */
+    }
+
+    @Test
+    public void json_is_deserialized_to_a_record() {
+        String json = gson.toJson(createTestBlog());
+        var blog = gson.fromJson(json, Blog.class);
+        System.out.println(blog);
+
+        // outputs:
+        // Blog[id=1, posts=[Post[blogId=1, author=Janne, title=This is title, content=This is content]]]
+    }
+
+    private static Blog createTestBlog() {
+        var post = new Post(1, "Janne", "This is title", "This is content");
+        var blog = new Blog(1, List.of(post));
+        return blog;
+    }
+}
+```
+
+## Thoughts after the experiment?
+
+Records provide a compact syntax for declaring simple, immutable data carrier classes. Frameworks can easily translate values from and to json (or any other serialization format).
+
+If you need to create instances of records by hand, you might get into trouble, because sometimes the data carrier classes contain a lot of fields.
+It's easy to mistype the order of two parameters with the same type. Proper typing helps mitigate the problem. In addition you can use the builder pattern. 
+For example, [Github - RecordBuilder](https://github.com/Randgalt/record-builder) seems very nice and allows you to write:
+
+```
+@RecordBuilder
+record Point(x: int, y: int) {}
+```
+
+After setting up the RecordBuilder annotation processor, it's easier to build instances, create copies and so on: 
+```
+var p1 = Point.builder().x(xCoord).y(yCoord).build();
+```
+ 
+To conclude, I would say that Java 14 version of records is a definitely a nice step forward.
+ 1. they reduce boilerplate from data carrier classes
+ 2. data is modeled as data
+
+The fact that data is modeled as data is even more important then reducing boilerplate, because it allows tools and
```

Comment:
'In fact, that latter is even more important than the former, because it allows tools and future versions of Java to support cool features such as pattern matching!'